### PR TITLE
feat: Add support for Action Items

### LINKED ISSUE #1595: Create support for `Action Items`
**Is your feature request related to a problem? Please describe.**

We need the ability to assign people to Action Items that can be tied to: 

1. A person at an event
2. A person independent of an event
3. A person and a non-event

This would be used to create a simpler, more universal alternative to Event Project Tasks on which a wider range of services could be built.

**Describe the solution you'd like**

To implement this we’ll need an Action Item related collection with the following fields:

1. ID
2. Assigned To (UserID)
3. Assigned By (UserID)
4. Topic ID 
5. Notes Prior to Completion
6. Notes on Completion
7. Assignment Date
8. Due Date (Should default to a week upon creation)
9. Completion Date (Should default to a week upon creation)
10. Completed (True/False)
11. Event ID -  Null if not related to an event (Generic Checklist)
12. updatedBy - See https://github.com/PalisadoesFoundation/talawa-api/pull/1542
13. updatedAt - See https://github.com/PalisadoesFoundation/talawa-api/pull/1542
14. createdBy - See https://github.com/PalisadoesFoundation/talawa-api/pull/1542
15. createdAt - See https://github.com/PalisadoesFoundation/talawa-api/pull/1542

You will need to create CRUD mutations and queries for these fields

The supporting Topic collection would have these fields:

1. ID
2. Topic - A subject area that would be of interest to the organization. (eg: Youth Group, Membership, Care etc.)
3. OrgID
4. Disabled/Enabled (True / False)
5. updatedBy - See https://github.com/PalisadoesFoundation/talawa-api/pull/1542
6. updatedAt - See https://github.com/PalisadoesFoundation/talawa-api/pull/1542
7. createdBy - See https://github.com/PalisadoesFoundation/talawa-api/pull/1542
8. createdAt - See https://github.com/PalisadoesFoundation/talawa-api/pull/1542

You will need to create CRU (no deletions) mutations and queries for these fields. The organization may want to disable a Topic so that it is no longer selectable, but still visible for pre-existing Action Items.

It is very possible that Action Items could be used beyond Events, and Generic Checklists. For example, they could be used to track activities related to a venue. Designing some way to make an action item usable for other purposes would be very useful. This would mean that the EventID field would be replaced with something else. Think of ways that this could be implemented simply.

**Describe alternatives you've considered**

- Event Project Tasks (See above)

**Approach to be followed (optional)**
- See above

**Additional context**
1. This is a prerequisite for our menu of GSoC 2024 project ideas
1. This issue would be best handled by someone who already has experience with updating our schema, mutations etc or can prove it in other scenarios. 
1. Please work with @xoldyckk on the details of this issue and the subsequent PR

**Potential internship candidates**
- Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by meetulr**
Fixes #1595 

**Did you add tests for your changes?**

Yes

**Summary**

This pull request introduces the following changes:
1. **Support for Categories:** Create, Read, and Update (CRU) operations have been added for Categories. This feature is accessible to superAdmins and orgAdmins.
2. **Support for Action Items:** Create, Read, Update, and Delete (CRUD) operations have been added for Action Items. This feature is accessible to superAdmins, orgAdmins, and eventAdmins.
3. **Category-Organization Relationship:** A two-way relationship has been established between Categories and Organizations. A default Category is automatically created and associated with each newly created Organization. Additionally, any new Categories created are automatically associated with the corresponding Organization.
4. **ActionItem-Event Relationship:** A two-way relationship has been established between Action Items and Events. If an Action Item is associated with an Event, it is automatically added to that Event. If an Action Item is deleted, it is automatically removed from the associated Event. Furthermore, if an Event is deleted, all associated Action Items are also deleted.
5. **Organization Cascade Delete Functionality:** Upon deletion of an Organization, all Categories associated with the Organization, as well as all Action Items associated with those Categories, are automatically deleted.
6. **Comprehensive Testing:** Tests have been added to ensure full coverage for all newly added or modified files.

**Does this PR introduce a breaking change?**

No

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by meetulr** [comment]
I've had fun implementing this. Let me know for any changes that are required, or anything I might have missed.
Thanks.

**Comment by palisadoes** [comment]
Just to confirm, did you implement Action Items related to Events and to Users independent of events?

**Comment by meetulr** [comment]
> Just to confirm, did you implement Action Items related to Events and to Users independent of events?

I didn't restrict the assigment of an action item to specific users, that means it could be assigned to anyone, regardless them being involved with the event or not.

**Comment by palisadoes** [comment]
- @xoldyckk
- @sumitra19jha
- @DMills27
- @xsh2047

Please review

**Comment by palisadoes** [comment]
Please fix the conflicting file

**Comment by meetulr** [comment]
Could this be reviewed please...
Sorry for asking like this😅.

**Comment by palisadoes** [comment]
Reviewers:

- @xoldyckk
- @sumitra19jha
- @DMills27
- @xsh2047
- @JamaicanFriedChicken 

This is ready for review.

**Comment by meetulr** [comment]
Resolved the conflicts.
Requesting review!😅

**Comment by sumitra19jha** [comment]
You have done changes in so many files. I will like too see a video recording as well of the working prototype in local.

**Comment by meetulr** [comment]
> You have done changes in so many files. I will like too see a video recording as well of the working prototype in local.

Almost all of the files are new, as it's a new feature.
The frontend for this has not yet been implemented.
Do you mean a demo in sandbox?
I have written tests for all the files too, and they're covering everything, and passing.

Shall I create an issue in admin for this? @palisadoes

**Comment by palisadoes** [comment]
@meetulr 

Yes prototype Admin screens for evaluation would be very helpful.

**Comment by palisadoes** [comment]
We should merge this prior to the Talawa Admin prototype after a review based on best practices. The admin prototype will highlight any broken functionality

Please review based on this assessment:

1. @xoldyckk
1. @sumitra19jha
1. @DMills27
1. @xsh2047
1. @JamaicanFriedChicken

**Comment by meetulr** [comment]
> @meetulr
> 
> Yes prototype Admin screens for evaluation would be very helpful.

Alright! I'll soon open an issue in admin for action items with my proposed solution, and if approved, I'll proceed with the implementation.

**Comment by meetulr** [comment]
> > @meetulr
> > Yes prototype Admin screens for evaluation would be very helpful.
> 
> Alright! I'll soon open an issue in admin for action items with my proposed solution, and if approved, I'll proceed with the implementation right away.

File: src/models/ActionItem.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import type { InterfaceEvent } from "./Event";
+import type { InterfaceCategory } from "./Category";
+
+/**
+ * This is an interface that represents a database(MongoDB) document for ActionItem.
+ */
+
+export interface InterfaceActionItem {
+  _id: Types.ObjectId;
+  assignedTo: PopulatedDoc<InterfaceUser & Document>;
```
Comment (xoldyckk) [inline src/models/ActionItem.ts:13]: rename to `assigneeId`

File: src/models/ActionItem.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  assignedBy: PopulatedDoc<InterfaceUser & Document>;
```
Comment (xoldyckk) [inline src/models/ActionItem.ts:14]: rename to `assignerId`

File: src/models/ActionItem.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  categoryId: PopulatedDoc<InterfaceCategory & Document>;
+  preCompletionNotes: string;
+  postCompletionNotes: string;
+  assignmentDate: Date;
+  dueDate: Date;
+  completionDate: Date;
+  completed: boolean;
```
Comment (xoldyckk) [inline src/models/ActionItem.ts:21]: rename to `isCompleted`

File: src/models/ActionItem.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  eventId: PopulatedDoc<InterfaceEvent & Document>;
+  createdBy: PopulatedDoc<InterfaceUser & Document>;
```
Comment (xoldyckk) [inline src/models/ActionItem.ts:23]: rename to `creatorId`

File: src/models/ActionItem.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ * @param dueDate - Due date.
+ * @param completionDate - Completion date.
+ * @param completed - Whether the ActionItem has been completed.
+ * @param eventId - Event to which the ActionItem is related, refer to the `Event` model.
+ * @param createdBy - User who created the ActionItem, refer to the `User` model.
+ * @param createdAt - Timestamp when the ActionItem was created.
+ * @param updatedAt - Timestamp when the ActionItem was last updated.
+ */
+
+const actionItemSchema = new Schema(
```
Comment (xoldyckk) [inline src/models/ActionItem.ts:45]: confirm which fields are non nullable and add `required: true` field to them

File: src/models/Category.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+ * @param orgId - Organization the category belongs to, refer to the `Organization` model.
+ * @param disabled - Whether category is disabled or not.
+ * @param createdBy - Task creator, refer to `User` model.
+ * @param createdAt - Time stamp of data creation.
+ * @param updatedAt - Time stamp of data updation.
+ */
+
+const categorySchema = new Schema(
+  {
+    category: {
```
Comment (xoldyckk) [inline src/models/Category.ts:32]: rename to `name`

File: src/models/Category.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      type: String,
+      required: true,
+    },
+    orgId: {
```
Comment (xoldyckk) [inline src/models/Category.ts:36]: rename to `organizationId`

File: src/models/Category.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      type: Schema.Types.ObjectId,
+      ref: "Organization",
+      required: true,
+    },
+    disabled: {
+      type: Boolean,
+      default: false,
```
Comment (xoldyckk) [inline src/models/Category.ts:43]: rename to `isDisabled`, `required: true`

File: src/models/Category.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    },
+    createdBy: {
```
Comment (xoldyckk) [inline src/models/Category.ts:45]: rename to `creatorId`

File: src/models/Category.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
+import type { PopulatedDoc, Types, Document, Model } from "mongoose";
+import { Schema, model, models } from "mongoose";
+import type { InterfaceUser } from "./User";
+import type { InterfaceOrganization } from "./Organization";
+
+/**
+ * This is an interface that represents a database(MongoDB) document for Category.
+ */
+
+export interface InterfaceCategory {
```
Comment (xoldyckk) [inline src/models/Category.ts:10]: Category is a very generic term, should be named something specific

File: src/models/ActionItem.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import { Schema, model, models } from "mongoose";
+import type { InterfaceUser } from "./User";
+import type { InterfaceEvent } from "./Event";
+import type { InterfaceCategory } from "./Category";
+
+/**
+ * This is an interface that represents a database(MongoDB) document for ActionItem.
+ */
+
+export interface InterfaceActionItem {
```
Comment (xoldyckk) [inline src/models/ActionItem.ts:11]: should probably be named `UserAction`?

File: src/models/Event.ts [hunk 12]
```diff
@@ -137,6 +140,12 @@ const eventSchema = new Schema(
       ref: "User",
       required: true,
     },
+    actionItems: [
```
Comment (xoldyckk) [inline src/models/Event.ts:143]: get rid of this, relationship to an event is already present on `ActionItem` using `eventId` field

File: src/models/Organization.ts [hunk 13]
```diff
@@ -84,6 +87,12 @@ const organizationSchema = new Schema(
       ref: "User",
       required: true,
     },
+    actionCategories: [
```
Comment (xoldyckk) [inline src/models/Organization.ts:90]: get rid of this, relationship to an organization is already present on `Category` using `organizationId` field

File: src/models/Category.ts [hunk 14]
```diff
@@ -0,0 +1,60 @@
+import type { PopulatedDoc, Types, Document, Model } from "mongoose";
```
Comment (xoldyckk) [inline src/models/Category.ts:1]: create a compound unique constraint on `category.organzationId` and `category.name`(category.category here)

**Comment by meetulr** [comment]
Thank you! I'll make the necessary changes.

**Comment by meetulr** [comment]
@palisadoes Shall I rename Category to ActionCategory, and ActionItem to UserAction?

**Comment by palisadoes** [comment]
The category should be renamed. Action Item should remain

**Comment by palisadoes** [comment]
1. The category should be renamed. Action Item Category to prevent any ambiguity
1. Action Item should remain

**Comment by palisadoes** [comment]
@meetulr 

1. Let us know when this is ready to be reviewed.
2. Mark the comments that have been rectified as being resolved

**Comment by palisadoes** [comment]
This is ready for review. Please take a look.

1. @xoldyckk
1. @sumitra19jha
1. @DMills27
1. @xsh2047
1. @JamaicanFriedChicken
1. @EshaanAgg

**Comment by meetulr** [comment]
This is a sandbox demo of the mutations and queries I made:

1. **For Action Item Categories**:

https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/02f61ddb-e424-4773-9493-d43513cb0f72

**Comment by meetulr** [comment]
This is a sandbox demo for the changes I made:

1. **For Action Item Categories**:

https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/4c5ddd2d-bba3-49d5-9096-fbccf2301f12

**Comment by SiddheshKukade** [comment]
I'll test it out and let you know.

**Comment by meetulr** [comment]
> This is a sandbox demo for the changes I made:
> 
> 1. **For Action Item Categories**:
> 
>  actionItemCategory.mp4 
> 2. **For Action Items**:
> 
>  actionItems.mp4

@xoldyckk 
@sumitra19jha 
@EshaanAgg 
@DMills27 
@JamaicanFriedChicken 

The frontend for this is now in development, any changes made/required here would have an impact there too.
Please take a look!😅
Thank you.

**Comment by meetulr** [comment]
> I'll test it out and let you know.

Thank you!

File: src/models/ActionItem.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    },
+    assignmentDate: {
+      type: Date,
+      required: true,
+      default: Date.now(),
+    },
+    dueDate: {
+      type: Date,
+      required: true,
+      default: Date.now() + 7 * 24 * 60 * 60 * 1000,
```
Comment (EshaanAgg) [inline src/models/ActionItem.ts:76]: Due to use constants inline such as `7 * 24 * ....`. Define it in `constants.ts` and then import and use the same here.

**Review by EshaanAgg — CHANGES_REQUESTED** [review]
Just a minor suggestion.

**Comment by meetulr** [comment]
> Just a minor suggestion.

Thank you! I've made the change, and I'll keep that in mind.

**Comment by palisadoes** [comment]
@EshaanAgg please review the updates

**Review by EshaanAgg — APPROVED** [review]

**Review by palisadoes — APPROVED** [review]

**Pull request merged**